### ISSUE TITLE
mark a specific package from a specific channel as bad or mark a package to always come from a specific channel

### ISSUE THREAD

**Issue by JanSchulz** [conda/conda#2067]
In some cases it could be good to not get a package from a specific channel (e.g. when conda-forge brings out a new matplotlib and the default packages has a faulty one or so...). See https://github.com/conda-forge/conda-forge.github.io/issues/22 for the disccusion.

There could be two ways to implment that:
* marking packages as bad ("channel/package version") -> everything which matches that is ignored
* markign packages as good: ("channel/package") -> every package with that name from a different channel gets ignored. 

An idea for the "mark bad" would be this:

* add a new method in `Resolve` ([code](https://github.com/conda/conda/blob/19c2d2e2393d6dd89cb8f051858da27f5cc8a78f/conda/resolve.py#L187) to remove all packages from the index, which have a specific name, version and channel 
* add a file named `<env-prefix>/conda-meta/marked-bad` which lists `channel/package version`, and then in https://github.com/conda/conda/blob/05366e9a259e0eb5301f0cb822781fadac100cad/conda/plan.py#L402 interate over this file and remove items by calling the new method on `r`

Alternative:
* add the removal directly to `get_index` if there is a `prefix/conda-meta/marked-bad` [code](https://github.com/conda/conda/blob/d3a76d3a0afbd31358464216a423806bb8e97e4f/conda/api.py#L44) 

Implementing the opposite would be almost the same, only you would give a spec ("channel/package") and then all packages with that name but not from the channel (`channel not in info["channel"]`) would be removed.

**Comment by groutr** [comment]
I guess a third option for implementation is a combination of blacklists and whitelists.

**Comment by mcg1969** [comment]
Seems to me a `marked-bad` option provides little additional benefit over proper channe prioritization plus the ability to fix the channel a particular package is coming from.

**Comment by JanSchulz** [comment]
IMO the key is individual package prioritizing, so that one could overwrite a specific package to always (not) come from a channel.

Only prioritize channels is IMO not fixing this special use case, as e.g. I want packages from the conda-forge channel as long as they don't exist in the main channel (because I don't want to have funny bugs which might exist due to incompatibilities), but for this specific package, I want to prioritize the one in conda-forge (because e.g. the package in the main channel is currently broken).

**Comment by mcg1969** [comment]
Right. We have a standard set of channel priorities, but allow them to be overridden on a per-channel basis.

**Comment by JanSchulz** [comment]
@mcg1969 Do you mean that there is already such a system? I only found https://github.com/conda/conda/issues/789 and this speaks about a future solution and also said that per package is not in the scope: https://github.com/conda/conda/issues/789#issuecomment-96019715.

**Comment by mcg1969** [comment]
Sorry to confuse you, no, there is not. I'm just brainstorming here. But I do think we're going to get this in the next major release (the one with my upgraded solver).

**Comment by kalefranz** [comment]
Rework of config.py is slated for `4.2` and this could probably take advantage of that.

**Comment by janschulz** [comment]
blacklisting packages probably better discussed further in #2410

**Comment by kalefranz** [comment]
I know this is fairly high priority, but it's not making it in 4.2, which will have a canary release on Thursday.

**Comment by kalefranz** [comment]
The part of this feature request

> mark a package to always come from a specific channel

is completed with conda 4.4, which will be rolling through the canary process over the next month.

**Comment by jakirkham** [comment]
Has there been any discussion of how to add this feature to a `conda-build` recipe?
